### PR TITLE
Convert model:// to Fuel URLs instead of absolute paths

### LINKED ISSUE #77: Stop hardcoding absolute paths into downloaded models
When `ign-fuel-tools` finds a `model://` within a file that it's downloading, it replaces that with the full path on disk where the file is being downloaded to. This is done by the [LocalCachePrivate::FixPaths*](https://github.com/ignitionrobotics/ign-fuel-tools/blob/master/src/LocalCache.cc) functions.

This is undesirable because it makes those models no longer be relocatable after they've been downloaded.

Moreover, in general, modifying a resource while downloading it makes for poor user experience, because the user is expecting to get the resource as it is on the server. Downloading a resource from the web UI or making direct REST calls to the server doesn't modify resources, and users are using these to circumvent this behaviour (among others) ([pit_crew](https://github.com/osrf/traffic_editor/blob/master/building_map_tools/pit_crew/pit_crew.py), [untouched](https://gist.github.com/chapulina/8f70c7298b2cba72460f6c30789c9179)).

There is a reason for the current behaviour though. Initially Ignition Gazebo didn't support `model://` URIs. And even now that it's about to support them as first-class citizens (https://github.com/ignitionrobotics/ign-gazebo/pull/172), that doesn't map well to models in the Fuel cache, because:

* The cache directory structure, `<model name>/version/model.sdf`, doesn't match the `<model name>/model.sdf` pattern used with the environment variable.
* Most of the models in the [OpenRobotics](https://app.ignitionrobotics.org/OpenRobotics) org have been transferred from [gazebo_models](https://github.com/osrf/gazebo_models), renamed on the process, but the URIs inside the SDF file remained the same. This is an issue with the resources though, not the library.

---

### Relative paths

Modify all models in Fuel that are using `model://` to use relative paths instead.

Problems with this approach:
* We do support `model://` though (actually any `scheme://`), so users may upload that and it has to work without overwriting.
* Relative paths don't support sharing resources across several models.

### Fuel cache + model://

Make it possible to find a resource on the cache, such as:

`~/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/Post\ office/2/meshes/post_office.dae`

from

`model://Post office/meshes/post_office.dae`

The main challenge is, what if multiple owners have models named `Post office`, which one do we get?

### PR DESCRIPTION
**Pull request by chapulina**
Builds on top of #84 .

Resolves #77.

## Summary

Without this PR, `model://` URIs like this:

`<uri>model://Radio/meshes/Radio.dae</uri>`

Would become absolute paths like this:

`<uri>/home/chapulina/.ignition/fuel/fuel.ignitionrobotics.org/OpenRobotics/models/Radio/4/meshes/Radio.dae</uri>`

With this PR, it becomes a Fuel URL like this:

`<uri>https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/Radio/4/files/meshes/Radio.dae</uri>`

## Resources across models

It's common for one model to use resources from other models to avoid duplication of resources. For example, the `Radio` model could use `model://Phone/meshes/antenna,dae` from the `Phone` model (assuming both.

For this to work on Blueprint and Citadel, `Radio` will need to use the full URL of the phone's antenna, i.e. 

`https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/Phone/4/files/meshes/antenna.dae`.

From `Dome` onwards, the same owner will be assumed for both models, and

`model://Phone/meshes/antenna,dae`

will be converted to the URL above automatically.

The reason not to do this for Blueprint is that there's a non-trivial number of models on Fuel right now whose Fuel name doesn't match the `model://` URI. These would look like different models which don't exist. For example, the [Construction Cone](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Construction%20Cone) uses `model://construction_cone`.

Right now, these models load fine, but a debug message is written:

```
[Dbg] [LocalCache.cc:583] Model [Construction Cone] loading resource from another model, named [construction_cone]. On Blueprint and Citadel, [construction_cone] is ignored. From Dome, [Construction Cone] will be used. If [construction_cone] is not a model belonging to owner [OpenRobotics], fix your SDF file!
```

### CODE REVIEW

File: src/LocalCache.cc [hunk 1]
```diff
@@ -556,34 +559,54 @@ bool LocalCachePrivate::FixPaths(const std::string &_modelVersionedDir)
...
+  auto resourceName = oldUri.substr(prefix.size(), firstSlash - prefix.size());
+
+  if (resourceName != _id.Name())
+  {
+    // On Blueprint and Citadel, just warn the user
+    // From Dome, use the name on the URI (resourceName) and assume the same
+    // owner
+    igndbg << "Model [" << _id.Name()
+           << "] loading resource from another model, named [" << resourceName
+           << "]. On Blueprint and Citadel, [" << resourceName << "] is "
```
Comment (nkoenig) [inline src/LocalCache.cc:585]: Is this comment is specific to `ign-gazebo`? If so, maybe you could add some language to help clarify.

File: src/LocalCache.cc [hunk 2]
```diff
@@ -556,34 +559,54 @@ bool LocalCachePrivate::FixPaths(const std::string &_modelVersionedDir)
...
+           << "ignored. From Dome, [" << _id.Name() << "] will be used. If ["
+           << resourceName << "] is not a model belonging to owner ["
+           << _id.Owner() << "], fix your SDF file!" << std::endl;
   }
+
+  auto filePath = oldUri.substr(firstSlash);
+
+  auto fuelUrl =
```
Comment (nkoenig) [inline src/LocalCache.cc:593]: Can you add a comment that describes what this is doing?

Comment (chapulina) [reply src/LocalCache.cc:596]: 05bf7412be60ce94bd612371ebfcaf7a2197ca20

File: src/LocalCache.cc (hunk 1 above)
Comment (chapulina) [reply src/LocalCache.cc]: Not really, this should apply to Fuel in general. I added specific Fuel versions in 05bf7412be60ce94bd612371ebfcaf7a2197ca20, let me know if you were thinking of something else.

**Comment by chapulina** [comment]
> Do you want to add a test, or save that for later?

The tests in `src/FuelClient_TEST.cc` were updated to reflect the changes in this PR. They check that resource tags now start with `https://`. I didn't want to check the entire URL in case the model changes.

**Comment by nkoenig** [comment]
I'm seeing a lot of errors with models in Subt. For example:

```
[Err] [Ogre2MeshFactory.cc:430] Cannot load null mesh [https://fuel.ignitionrobotics.org/1.0/OpenRobotics/models/X1 Config 1/15/files/meshes/wheel.dae]
```

**Comment by chapulina** [comment]
> I'm seeing a lot of errors with models in Subt. For example:

Those work for me. Are you running the latest commit of `ign-gazebo2`?

**Comment by nkoenig** [comment]
> > I'm seeing a lot of errors with models in Subt. For example:
> 
> Those work for me. Are you running the latest commit of `ign-gazebo2`?

That fixed the problem, however this change breaks `gz3d`. I'll have to fixed that library first.

**Comment by chapulina** [comment]
> this change breaks gz3d. I'll have to fixed that library first.

Oh I hadn't thought of that. Maybe it's safer to target this at Dome to preserve backwards compatibility. Let me know how gz3d goes.

**Comment by nkoenig** [comment]
> > this change breaks gz3d. I'll have to fixed that library first.
> 
> Oh I hadn't thought of that. Maybe it's safer to target this at Dome to preserve backwards compatibility. Let me know how gz3d goes.

Here is the patch. I'll approve this PR so that we can eventually make new docker images for SubT.

**Pull request merged**